src/backend/controllers/mealsController.js

```diff
@@ -0,0 +1,164 @@
+  try {
+    const mealId = request.params.id;
+    const meal = await knex("meal").where("id", mealId).first();
+    if (!meal) {
+      return response.status(404).json({ error: "meal not found" });
+    }
+    response.json(meal);
+  } catch (error) {
+    res.status(500).json({ error: "Internal server error" });
+  }
+};
+// /api/meals/:id PUT	Updates the meal by id
+const updateMeal = async (req, res) => {
+  const mealId = req.params.id;
+  const { title, description, location, when, max_reservations, price } =
+    req.body;
+  const { error, value } = validateMeal(req.body);
+  if (error) {
+    return res.status(400).json({ error: error.details[0].message });
+  }
+  try {
+    // Update the meal in the database using Knex
+    await knex("meal").where("id", mealId).update({
+      title,
+      description,
+      location,
+      when,
+      max_reservations,
+      price,
+    });
+
+    res.sendStatus(204); // Successful update, no content
+  } catch (error) {
+    res.sendStatus(500); // Internal server error
+  }
+};
+// /api/meals/:id	DELETE	Deletes the meal by id
+const deleteMeal = async (req, res) => {
+  const mealId = req.params.id;
+  try {
+    // checking whether mealis exits
+    const mealExists = await knex("meal").where("id", mealId).first();
+    if (!mealExists) {
+      return res.status(404).send("Meal not found.Please provide a valid id");
+    }
+    // delete the meal in the database using Knex
+    await knex("meal").where("id", mealId).del();
+    res.sendStatus(200);
+  } catch (error) {
+    res.sendStatus(500); // Internal server error
+  }
+};
+// HW-Week3  GET api/meals route to add query parameters
+
+const getMealsByQuery = async (req, res) => {
+  try {
+    // extract query parameters from the request
+    const {
+      maxPrice,
+      availableReservations,
+      title,
+      dateAfter,
+      dateBefore,
+      limit,
+      sortKey,
+      sortDir,
+    } = req.query;
+    //initial query using knex
+    let query = knex("meal");
+    if (maxPrice) {
+      query = query.where("price", "<=", parseFloat(maxPrice));
+    }
+    if (availableReservations === "true") {
+      query = query.where(
+        "max_reservations",
+        ">",
+        knex.raw("COUNT(reservation.id)")
+      );
+      query = query.leftJoin("reservation", "meal.id", "reservation.,meal_id");
+      query = query.groupBy("meal_id");
+    }
+    if (title) {
+      query = query.where("title", "like", "%${title}");
+    }
+    if (dateAfter) {
+      query = query.where("when", ">", dateAfter);
+    }
+    if (dateBefore) {
+      query = query.where("when", "<", dateBefore);
+    }
+    if (limit) {
+      query = query.limit(parseInt(limit));
+    }
+    if (sortKey) {
+      const allowedSortKeys = ["when", "max_reservations", "price"];
+      if (allowedSortKeys.includes(sortKey)) {
+        const direction = sortDir === "desc" ? "desc" : "asc";
+        query = query.orderBy(sortKey, direction);
```

Comment:
I would prefer using knex.js parameterized queries. If there is any doubt, check out warmup review, you will have a hint to refactor the code.